File: modules/project/src/main/java/com/falsepattern/zigbrains/project/ide/project/ZigProjectConfigurable.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.project.ide.project;

import com.falsepattern.zigbrains.common.SubConfigurable;
import com.falsepattern.zigbrains.common.util.dsl.JavaPanel;
import com.falsepattern.zigbrains.project.openapi.components.ZigProjectSettingsService;
import com.falsepattern.zigbrains.zig.lsp.ZLSStartupActivity;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Disposer;
import lombok.val;
import org.jetbrains.annotations.NotNull;

public class ZigProjectConfigurable implements SubConfigurable {
    private ZigProjectSettingsPanel settingsPanel;

    private final Project project;

    public ZigProjectConfigurable(@NotNull Project project) {
        this.project = project;
    }

    @Override
    public void createComponent(JavaPanel panel) {
        settingsPanel = new ZigProjectSettingsPanel();
        settingsPanel.attachPanelTo(panel);
    }

    @Override
    public boolean isModified() {
        return ZigProjectSettingsService.getInstance(project).isModified(settingsPanel.getData());
    }

    @Override
    public void apply() throws ConfigurationException {
        val service = ZigProjectSettingsService.getInstance(project);
        val data = settingsPanel.getData();
        val modified = service.isModified(data);
        service.loadState(data);
        if (modified) {
            ZLSStartupActivity.initZLS(project);
        }
    }

    @Override
    public void reset() {
        val zigSettings = ZigProjectSettingsService.getInstance(project);
        settingsPanel.setData(zigSettings.getState());
    }

    @Override
    public void disposeUIResources() {
        Disposer.dispose(settingsPanel);
        settingsPanel = null;
    }
}

File: core/src/main/kotlin/com/falsepattern/zigbrains/project/settings/ZigProjectConfigurable.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.project.settings

import com.falsepattern.zigbrains.shared.SubConfigurable
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Disposer
import com.intellij.ui.dsl.builder.Panel

class ZigProjectConfigurable(private val project: Project): SubConfigurable {
    private var settingsPanel: ZigProjectSettingsPanel? = null
    override fun createComponent(holder: ZigProjectConfigurationProvider.SettingsPanelHolder, panel: Panel): ZigProjectConfigurationProvider.SettingsPanel {
        settingsPanel?.let { Disposer.dispose(it) }
        val sp = ZigProjectSettingsPanel(holder, project).apply { attach(panel) }.also { Disposer.register(this, it) }
        settingsPanel = sp
        return sp
    }

    override fun isModified(): Boolean {
        return project.zigProjectSettings.isModified(settingsPanel?.data ?: return false)
    }

    override fun apply() {
        val service = project.zigProjectSettings
        val data = settingsPanel?.data ?: return
        val modified = service.isModified(data)
        service.state = data
        if (modified) {
            ZigProjectConfigurationProvider.mainConfigChanged(project)
        }
    }

    override fun reset() {
        settingsPanel?.data = project.zigProjectSettings.state
    }

    override fun dispose() {
        settingsPanel = null
    }
}